### PR TITLE
auto generate extensions.go

### PR DESCRIPTION
**Pull request by zirain**
fixes: https://github.com/envoyproxy/gateway/issues/1116

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -87,6 +87,7 @@ require (
 	github.com/sirupsen/logrus v1.8.1 // indirect
 	github.com/tsaarni/x500dn v0.0.0-20210331182804-14283c7f5a16 // indirect
 	github.com/xlab/treeprint v1.1.0 // indirect
+	go.opentelemetry.io/proto/otlp v0.19.0 // indirect
```
Comment (arkodg) [inline go.mod:90]: why is this added ?

File: tools/make/golang.mk [hunk 2]
```diff
@@ -32,7 +32,7 @@ go.build.%:
 
 # Build the envoy-gateway binaries in the hosted platforms.
 .PHONY: go.build
-go.build: $(addprefix go.build., $(addprefix $(PLATFORM)., $(BINS)))
+go.build: go.generate $(addprefix go.build., $(addprefix $(PLATFORM)., $(BINS)))
```
Comment (arkodg) [inline tools/make/golang.mk:35]: this shouldnt be dependancy imo

File: tools/make/golang.mk [hunk 3]
```diff
@@ -67,6 +67,11 @@ go.mod.lint:
 		$(call log, "Go module looks clean!"); \
    	fi
 
+.PHONY: go.generate
+go.generate: ## Generate code from templates
```
Comment (arkodg) [inline tools/make/golang.mk:71]: needs to be linked to `make generate`

File: internal/xds/extensions/extensions.go [hunk 4]
```diff
@@ -3,143 +3,17 @@
...
-)
+// nolint: lll
+//
+//go:generate sh -c "echo '// Copyright Envoy Gateway Authors' > extensions.gen.go"
+//go:generate sh -c "echo '// SPDX-License-Identifier: Apache-2.0' >> extensions.gen.go"
+//go:generate sh -c "echo '// The full text of the Apache license is available in the LICENSE file at' >> extensions.gen.go"
+//go:generate sh -c "echo '// the root of the repo.\n' >> extensions.gen.go"
+//go:generate sh -c "echo '// GENERATED FILE -- DO NOT EDIT\n' >> extensions.gen.go"
+//go:generate sh -c "echo 'package extensions\n\nimport (' >> extensions.gen.go"
+//go:generate sh -c "go list github.com/envoyproxy/go-control-plane/... | grep 'v[2-9]' | grep -v /pkg/ | xargs -I{} echo '\t_ \"{}\"' >> extensions.gen.go"
```
Comment (arkodg) [inline internal/xds/extensions/extensions.go:18]: 🚀  thanks for automating this !

File: tools/make/golang.mk (hunk 2 above)
Comment (zirain) [reply tools/make/golang.mk:35]: you should run `go generate` before `go build`

File: tools/make/golang.mk (hunk 3 above)
Comment (zirain) [reply tools/make/golang.mk:71]: IMO, linked to gen-check is enough.

File: go.mod (hunk 1 above)
Comment (zirain) [reply go.mod:90]: because of `github.com/envoyproxy/go-control-plane/envoy/extensions/access_loggers/open_telemetry/v3`

File: tools/make/golang.mk (hunk 2 above)
Comment (arkodg) [reply tools/make/golang.mk:35]: we already generate a lot of go code today 
https://github.com/envoyproxy/gateway/blob/630cae98c9fb6e66f303ef9174876aa6d21186b6/tools/make/kube.mk#L32
I suggest we move this target into `common.mk` and also add a `go generate` step in there

File: go.mod (hunk 1 above)
Comment (arkodg) [reply go.mod:90]: ah thanks for the clarification !

File: tools/make/golang.mk (hunk 2 above)
Comment (zirain) [reply tools/make/golang.mk:35]: yeah, that make sense

Comment (zirain) [reply tools/make/golang.mk]: updated, PTAL

Comment (arkodg) [reply tools/make/golang.mk]: thanks, it looks good, can you confirm that `make geo-check` works with no hiccups ?
can you also update the print statement in https://github.com/envoyproxy/gateway/blob/630cae98c9fb6e66f303ef9174876aa6d21186b6/tools/make/lint.mk#L72 ?

File: tools/make/common.mk [hunk 5]
```diff
@@ -113,6 +113,9 @@ Options:
 endef
 export USAGE_OPTIONS
 
+.PHONY: generate
+generate: kube-generate go.generate
```
Comment (arkodg) [inline tools/make/common.mk:117]: can you add a helper string that will show up with `make help` - `## Generate go code from templates and tags`

File: tools/make/golang.mk (hunk 2 above)
Comment (zirain) [reply tools/make/golang.mk]: the output looks expected to me
```
make[1]: Entering directory '/root/go/src/github.com/envoyproxy/gateway'
===========> Running kube-generate ... 
tools/bin/controller-gen object:headerFile="/root/go/src/github.com/envoyproxy/gateway/tools/boilerplate/boilerplate.generatego.txt",year=2023 paths="{/root/go/src/github.com/envoyproxy/gateway/api/...,/root/go/src/github.com/envoyproxy/gateway/internal/ir/...,/root/go/src/github.com/envoyproxy/gateway/internal/gatewayapi/...}"
===========> Running go.generate ... 
go generate ./...
===========> Running manifests ... 
tools/bin/controller-gen rbac:roleName=envoy-gateway-role crd webhook paths="./..." output:crd:artifacts:config=charts/eg/crds/generated output:rbac:artifacts:config=charts/eg/templates/generated/rbac output:webhook:artifacts:config=charts/eg/templates/generated/webhook
===========> Running gen-check ... 
make[1]: Leaving directory '/root/go/src/github.com/envoyproxy/gateway'
```

File: tools/make/common.mk (hunk 5 above)
Comment (zirain) [reply tools/make/common.mk:118]: done

**Review by arkodg — APPROVED** [review]
LGTM thanks for ensuring there is no drift in the future !
